kernel/queue.c

```diff
@@ -207,19 +207,23 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
 
 	event.state = K_POLL_STATE_NOT_READY;
 
-	err = k_poll(&event, 1, timeout);
-	if (err) {
-		return NULL;
-	}
+	do {
+		err = k_poll(&event, 1, timeout);
```

Comment:
I think you need to move the "event.state = K_POLL_STATE_NOT_READY;" into the loop so that it gets set every time before you call k_poll. That's at least how Ben instructed me to use the API when I originally took it into use in our Bluetooth code (which also calls it in a loop).

Reply:
Agreed, as stated in here: https://www.zephyrproject.org/doc/kernel/other/polling.html#using-k-poll